File: src/0033.Search_in_Rotated_Sorted_Array-Array/search_in_rotated_sorted_array.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution{
public:
    int search(vector<int>& nums, int target){
        int mid;
        int ll = 0;
        int rr = nums.size()-1;
        while(ll < rr){
            mid = (ll + rr)/2;
            if(nums[0] <= nums[mid] && (target < nums[0] || target > nums[mid])){
                ll = mid+1;
            }
            else if(nums[0] > nums[mid] && (target < nums[0] && target > nums[mid])){
                ll = mid+1;
            }
            else{
                rr = mid;
            }
        }
        return (ll == rr && nums[ll] == target) ? ll : -1;
    }
};

class Solution2 {
public:
    int search(vector<int>& nums, int target) {
        int ll = 0;
        int rr = nums.size() - 1;
        while(ll <= rr){
            int mid = ll + (rr - ll)/2;
            if(nums[mid] == target) return mid;
            if(nums[ll] <= nums[mid]){
                if(nums[ll] > target || nums[mid] < target){
                    ll = mid + 1;
                }
                else{
                    rr = mid - 1;
                }
            }
            else{
                if(target < nums[ll] && target > nums[mid]){
                    ll = mid + 1;
                }
                else{
                    rr = mid - 1;
                }
            }
        }
        return -1;
    }
};

int main()
{
    vector<int> nums = {4,5,6,7,0,1,2};
    int target = 0;
    Solution S;
    int res = S.search(nums, target);
    cout << res << endl;
    return 0;
}

File: src/0033.Search_in_Rotated_Sorted_Array-Array/search_in_rotated_sorted_array.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        while(left < right){
            int mid = left + (right - left + 1) / 2;
            if(nums[mid] < nums[right]){
                if(nums[mid] <= target && target <= nums[right]){
                    left = mid;
                } else{
                    right = mid - 1;
                }
            } else{
                if(nums[left] <= target && target <= nums[mid - 1]){
                    right = mid - 1;
                } else{
                    left = mid;
                }
            }
        }
        if(nums[left] == target){
            return left;
        }
        return -1;
    }
}